docs: document kubelet cgroup driver detection from the runtime (#42160)
kubernetes/website: content/en/docs/setup/production-environment/container-runtimes.md

the `cgroupDriver` field under `KubeletConfiguration`, kubeadm defaults it to `systemd`.
{{< /note >}}

> For Kubernetes v1.28 and later, with the `KubeletCgroupDriverFromCRI`

Reviewer: bit of a tldr issue I believe.. giving the instructions above for kubelet config driven and default if not configured.. then ending with 1.28 alpha feature gate info.

Suggest mentioning the feature gate above in direct association with the kubelet config steps / options. Something like unless the alpha level feature gate is set to automatically retrieve the appropriate cgroup driver from the runtime, additional configuration may be required...

Author: I'm a bit cautious on this ("advertising" too much) as this is an alpha feature and there are no released container runtime supporting this, yet. When we get to beta, sure. WDYT @sftim

Reviewer: good point...

Reviewer 2: We should document how Kubernetes behaves, even if the compatible runtimes don't exist on release day.

> In Kubernetes {{< skew currentVersion >}}, with the `KubeletCgroupDriverFromCRI`

Reviewer 3: Suppose we are in v1.29 now, the above line will be rendered into something like
"In Kubernetes 1.29, with the `KubeletCgroupDriverFromCRI`".
Is this statement still appropriate?
Suppose we are in v1.34 now and the feature gate has been GA'ed and removed,
the above line will be rendered into something like
"In Kubernetes 1.34, with the `KubeletCgroupDriverFromCRI`".
Is this statement still appropriate?

Clearly stating that this feature is in Alpha starting 1.28 is good enough.
Why do we use the `skew` shortcode the make this release number a moving target?

Author: I'm indifferent on this, can change it to v1.29. This (`In Kubernetes {{< skew currentVersion >}}`) was something that @sftim suggested in one of his earlier comments so I took it from there.

Reviewer 2: >  Suppose we are in v1.34 now and the feature gate has been GA'ed and removed,

When we change Kubernetes, we should remove the docs. If we don't trust ourselves to do that, I would focus on fixing that lack of trust.

If we could trust the release docs process, I think this change would be fine.

Author: As a contributor I'm inclined to agree with @sftim about the docs process here. Especially with a KEP involved, the process should ensure that the documentation is changed when the feature is moved to beta and later GA (or removed before graduating further). WDYT @tengqm?

Reviewer 2: Per https://github.com/kubernetes/website/pull/42160#issuecomment-1665322332, let's follow the path of least risk to the release, at least for now.

Author: K, I changed this now to `In Kubernetes v1.28, ...` (not to `Starting with v1.28 ...` as per https://github.com/kubernetes/website/pull/42160#discussion_r1274196798)

PTAL @sftim @tengqm

the `cgroupDriver` field under `KubeletConfiguration`, kubeadm defaults it to `systemd`.
{{< /note >}}

> {{< note >}}

Reviewer 2: I wouldn't use a note here. Use a plain paragraph.

Author: OK, made it a plain (non-note) paragraph, as suggested

> Starting with v1.28 and later, with KubeletCgroupDriverFromCRI feature gate

Reviewer 2: ```suggestion
For Kubernetes v1.28 and later, with the `KubeletCgroupDriverFromCRI`
[feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
```

Author: Thanks. Changed as suggested

> enabled and a container runtime that supports the RuntimeConfig CRI rpc,

Reviewer 2: ```suggestion
enabled and a container runtime that supports the `RuntimeConfig` CRI RPC,
```

Author: 👍 Changed as suggested

> kubelet will automatically detect the cgroup driver from the runtime and the
> cgroupDriver setting in kubelet configuration is ignored.

Reviewer 2: ```suggestion
the kubelet automatically detects the appropriate cgroup driver from the runtime,
and ignores the `cgroupDriver` setting within the kubelet configuration.
```

When using kubeadm, manually configure the
[cgroup driver for kubelet](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).

> Starting with v1.28 and later, you can enable automatic detection of the

Reviewer: ```suggestion
Starting with Kubernetes v1.28 and later, you can enable automatic detection of the
```

Reviewer 2: Aside, but also a bit relevant:
I'm slightly wary of saying “and later” about alpha features that we can't prove will graduate to stable.

Author: Fair point, have to agree on that. I now changed all of these to `In Kubernetes {{< skew currentVersion >}}...`

ok @sftim @mikebrow ?

Reviewer: given the concern over skew "and later" comments perhaps.. for Kubernetes v1.2.8, you can..

> In Kubernetes {{< skew currentVersion >}}, you can enable automatic detection of the

Reviewer 3: Same as above.

In addition to the above confusion, using this `skew` shortcode makes it difficult
to remove contents that are no longer relevant in the future.

When using kubeadm, manually configure the
[cgroup driver for kubelet](/docs/tasks/administer-cluster/kubeadm/configure-cgroup-driver/#configuring-the-kubelet-cgroup-driver).

> {{< note >}}
> Starting with v1.28 and later, kubelet cgroup driver does not need to be
> configured if the KubeletCgroupDriverFromCRI feature gate is enabled and a
> version of containerd that supports the RuntimeConfig CRI rpc is being used.
> {{< /note >}}

Reviewer 2: ```suggestion
Starting with v1.28 and later, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
```

or
```suggestion
In Kubernetes {{< skew currentVersion >}}, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
```

Author: Thx. Chose the first variant suggested above. I don't have strong opinions on this, but personally (as a user) like to easily get the information when some new feature was introduced.

cgroup driver configuration of the kubelet (usually done via kubeadm) and CRI-O
in sync.

> Starting with v1.28 and later, you can enable automatic detection of the

Reviewer: ```suggestion
Starting with Kubernetes v1.28 and later, you can enable automatic detection of the
```

Author: Fixed with `In Kubernetes {{< skew currentVersion >}}`

> In Kubernetes {{< skew currentVersion >}}, you can enable automatic detection of the

Reviewer 3: Same here.

cgroup driver configuration of the kubelet (usually done via kubeadm) and CRI-O
in sync.

> {{< note >}}
> Starting with v1.28 and later, kubelet cgroup driver does not need to be
> configured if the KubeletCgroupDriverFromCRI feature gate is enabled and a
> version of CRI-O that supports the RuntimeConfig CRI rpc is being used.
> {{< /note >}}

Reviewer 2: ```suggestion
Starting with v1.28 and later, you can enable automatic detection of the
cgroup driver as an alpha feature. See [systemd cgroup driver](#systemd-cgroup-driver)
for more details.
```

Author: Thx. Changed as suggested
